mentat/llm_api_handler.py

```diff
@@ -0,0 +1,214 @@
+        encoding = tiktoken.get_encoding("cl100k_base")
+    return len(encoding.encode(message, disallowed_special=())) + (
+        4 if full_message else 0
+    )
+
+
+def conversation_tokens(messages: list[ChatCompletionMessageParam], model: str):
+    """
+    Returns the number of tokens used by a full conversation.
+    Adapted from https://platform.openai.com/docs/guides/text-generation/managing-tokens
+    """
+    try:
+        encoding = tiktoken.encoding_for_model(model)
+    except KeyError:
+        encoding = tiktoken.get_encoding("cl100k_base")
+
+    num_tokens = 0
+    for message in messages:
+        # every message follows <im_start>{role/name}\n{content}<im_end>\n
+        num_tokens += 4
+        for key, value in message.items():
+            if not isinstance(value, str):
+                continue
+            num_tokens += len(encoding.encode(value))
+            if key == "name":  # if there's a name, the role is omitted
+                num_tokens -= 1  # role is always required and always 1 token
+    num_tokens += 2  # every reply is primed with <im_start>assistant
+    return num_tokens
+
+
+# TODO: This function should just get the prompt token count; it shouldn't be warning at all
+def get_prompt_token_count(
+    messages: list[ChatCompletionMessageParam], model: str
+) -> int:
+    session_context = SESSION_CONTEXT.get()
+    stream = session_context.stream
+
+    prompt_token_count = conversation_tokens(messages, model)
+
+    token_buffer = 500
+    context_size = model_context_size(model)
+    if context_size:
+        if prompt_token_count > context_size - token_buffer:
+            stream.send(
+                f"Warning: {model} has a maximum context length of {context_size}"
+                " tokens. Attempting to run anyway:",
+                color="yellow",
+            )
+    return prompt_token_count
+
+
+# TODO: These two functions should be a dictionary
+def model_context_size(model: str) -> Optional[int]:
+    if model == "gpt-4-1106-preview":
+        return 128000
+    elif "gpt-4" in model:
+        if "32k" in model:
+            return 32768
+        else:
+            return 8192
+    elif "gpt-3.5" in model:
+        if "16k" in model:
+            return 16385
+        else:
+            return 4097
+    elif "ada-002" in model:
+        return 8191
+    else:
+        return None
+
+
+def model_price_per_1000_tokens(model: str) -> Optional[tuple[float, float]]:
+    """Returns (input, output) cost per 1000 tokens in USD"""
+    if model == "gpt-4-1106-preview":
+        return (0.01, 0.03)
+    elif "gpt-4" in model:
+        if "32k" in model:
+            return (0.06, 0.12)
+        else:
+            return (0.03, 0.06)
+    elif "gpt-3.5" in model:
+        if "16k" in model:
+            return (0.003, 0.004)
+        else:
+            return (0.0015, 0.002)
+    elif "ada-002" in model:
+        return (0.0001, 0)
+    else:
+        return None
+
+
+class LlmApiHandler:
```

Comment:
@biobootloader suggested the same thing; like I said there, I think convention for PascalCase is not to capitalize abbreviations (otherwise we'd have LLMAPIHandler)